nixops/backends/__init__.py

```diff
@@ -212,7 +212,14 @@ def send_keys(self):
         for k, opts in self.get_keys().items():
             self.log("uploading key ‘{0}’...".format(k))
             tmp = self.depl.tempdir + "/key-" + self.name
-            f = open(tmp, "w+"); f.write(opts['text']); f.close()
+
+            # Ensure that keys.<key>.text XOR keys.<key>.filename
+            assert     ('text' in opts  or 'filename' in opts), "Specification error for key '%s': neither `keys.%s.text` nor `keys.%s.filename` was specified."  % (k, k, k)
+            assert not ('text' in opts and 'filename' in opts), "Specification error for key '%s': `keys.%s.text` and `keys.%s.filename` are mutually exclusive." % (k, k, k)
```

Comment:
Let's move this to Nix. For example like in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L539-L564